### PR TITLE
Split existing "Show all labels (including colliding labels)" into two separate new options

### LINKED ISSUE #41043: Label placement is broken in QGIS 3.16
Text placement is set to "Curved" along a line. When label is moved to the edge of the screen, then it's placement switches to horizontal, an sometimes label even overruns. But overrun is set to none. 

See the attached video.
There is a comparison between 3.12 and 3.16.

v3.12 behavior
https://user-images.githubusercontent.com/11345848/104806985-6ac31400-57ec-11eb-9ac0-cc350a0bcec9.mp4

v3.16 vehavior
https://user-images.githubusercontent.com/11345848/104806996-7adaf380-57ec-11eb-8b57-07af02d9bc6d.mp4

This breaks tiled export and exprort to partial map sheets.
I expect the label to be clipped to the edge of the screen (preferrable) or disappear.

### LINKED ISSUE #47482: Strict line label anchoring ignored if "show all labels" is enabled
### What is the bug or the crash?

When a line label uses a strict label anchor and "Show all labels for this layer" is enabled, the label will be shown at 50%of the line if the set position (plus the space that the label takes when placed there) is scrolled out of the canvas.

asd

As a user this is unexpected. There is no collision. The label should be drawn at exactly 20% of the entire line. If that point (resp. the label placed there) is outside the canvas, there should be simply no label.

### Steps to reproduce the issue

1. Have a line with an attribute for labeling
2. Set it to label with some label anchoring (start, center, middle, custom all lead to this), entire line clipping, strict placement
3. Enable "Show all labels for this layer"
4. Drag the canvas around so that the beginning of the line and the label are outside canvas
5. A label will be shown without rotation at 50% of the line

### Versions

3.22, master

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

I did not test this with polygons, if there is a similar option available, I suspect it might exhibit the same issue.

### PR DESCRIPTION
**Pull request by nyalldawson**
This setting was ambiguous in meaning -- it could be interpreted as either:

"I don't care about overlapping labels, use them when you have no other choice"

OR

"I want to see labels for EVERYTHING in this layer, and I don't care what you have to do to show me those"

This leads to situations like those discussed in #41043, where some users interpret the setting as the first meaning and get confused when eg. a curved label line layer starts showing horizontal placements for some features which couldn't otherwise be labeled (eg short lines).

So, split the setting into two new options (checkboxes) with explicit, clear distinction:

- "Allow inferior fallback placements": This mode permits features to fallback to worse placement options when there's no other choice, e.g. when a line is too short to fit a curved label text then the label may be placed horizontally just over the feature's center point.

- Multiple choices for overlapping label handling, with choices:

    - Never Overlap : never ever place overlapping labels for the layer, even if it means labels will be missing
    - Allow Overls if Required: if the label can't otherwise be placed, draw an overlapping label. This mode will cause the label to be moved a less ideal placement if possible, e.g. moving the label further from the center of a line or polygon, IF doing so will avoid overlapping labels. But if there's no other positions possible, then draw the label overlapping.
    - Always Allow Overlaps": It doesn't matter at all if the label overlaps other labels or obstacles, that's fine to do and the best
placement (e.g most central placement) should always be used even if an alternate further placement is possible which avoids overlaps entirely.

The combination of checking BOTH options WILL guarantee that all features in the layer are labelled, albeit with both overlapping labels and the degraded placements!

While the API is being completely reworked, I've also taken the opportunity to move these from being layer-wide settings to feature specific settings, so it's now possible to data - define their value on a feature-by-feature basis.

Fixes #41043
Fixes https://github.com/qgis/QGIS/issues/47482

### CODE REVIEW

**Comment by nyalldawson** [comment]
Opened early for a CI run only -- I'll add tests and adapt existing tests after

File: src/core/labeling/qgstextlabelfeature.h [hunk 1]
```diff
@@ -20,6 +20,7 @@
 #include "qgslabelfeature.h"
 #include "qgstextdocument.h"
 #include "qgstextmetrics.h"
+#include "qgspallabeling.h"
```
Comment (nirvn) [inline src/core/labeling/qgstextlabelfeature.h:23]: @nyalldawson , is that additional include required? Just checking.

File: src/core/labeling/qgsvectorlayerlabelprovider.h [hunk 2]
```diff
@@ -22,6 +22,7 @@
 #include "qgslabelingengine.h"
 #include "qgsrenderer.h"
 #include "qgstextrenderer.h"
+#include "qgspallabeling.h"
```
Comment (nirvn) [inline src/core/labeling/qgsvectorlayerlabelprovider.h:25]: Same here, required? (I'm starting to assume so seeing it twice :) )

File: src/gui/qgstextformatwidget.cpp [hunk 3]
```diff
@@ -1869,6 +1879,25 @@ void QgsTextFormatWidget::updateDataDefinedAlignment()
                                     || mCoordPointDDBtn->isActive() );
 }
 
+void QgsTextFormatWidget::overlapModeChanged()
+{
+  QString description;
+  switch ( static_cast< Qgis::LabelOverlapHandling >( mComboOverlapHandling->currentData().toInt() ) )
+  {
+    case Qgis::LabelOverlapHandling::PreventOverlap:
+      description = tr( "Overlapping labels will never be placed for the layer, even if it means some labels will be missing." );
```
Comment (nirvn) [inline src/gui/qgstextformatwidget.cpp:1888]: We could maybe add a tip along the lines of "To see unplaced labels, use the show unplaced labels toolbar action" too there. No strong opinions.

**Review by nirvn — APPROVED** [review]
@nyalldawson , nice cleanup work in there (i.e. class members initiated in header, enums moved to qgis, etc.).

File: src/core/labeling/qgstextlabelfeature.h (hunk 1 above)
Comment (nyalldawson) [reply src/core/labeling/qgstextlabelfeature.h:23]: Yes -- it was previously indirectly included through one of the other includes, now the includes are tightened up and the other file doesn't include it.

File: src/core/labeling/qgsvectorlayerlabelprovider.h (hunk 2 above)
Comment (nyalldawson) [reply src/core/labeling/qgsvectorlayerlabelprovider.h:25]: Yep, same reason. It's actually a tightening of includes and more heavy use of forward declrations.

File: src/gui/qgstextformatwidget.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/gui/qgstextformatwidget.cpp]: done

**Comment by kannes** [comment]
Very nice, thank you!

**Pull request merged**